src/s3.zig

```diff
@@ -0,0 +1,572 @@
+const bun = @import("root").bun;
+const picohttp = bun.picohttp;
+const std = @import("std");
+const DotEnv = @import("./env_loader.zig");
+
+const JSC = bun.JSC;
+const strings = bun.strings;
+
+pub const AWSCredentials = struct {
+    accessKeyId: []const u8,
+    secretAccessKey: []const u8,
+    region: []const u8,
+    endpoint: []const u8,
+    bucket: []const u8,
+    const log = bun.Output.scoped(.AWS, false);
+
+    const DateResult = struct {
+        // numeric representation of year, month and day (excluding time components)
+        numeric_day: u64,
+        date: []const u8,
+    };
+
+    fn getAMZDate(allocator: std.mem.Allocator) DateResult {
+        // We can also use Date.now() but would be slower and would add JSC dependency
+        // var buffer: [28]u8 = undefined;
+        // the code bellow is the same as new Date(Date.now()).toISOString()
+        // JSC.JSValue.getDateNowISOString(globalObject, &buffer);
+
+        // Create UTC timestamp
+        const secs: u64 = @intCast(@divFloor(std.time.milliTimestamp(), 1000));
+        const utc_seconds = std.time.epoch.EpochSeconds{ .secs = secs };
+        const utc_day = utc_seconds.getEpochDay();
+        const year_and_day = utc_day.calculateYearDay();
+        const month_and_day = year_and_day.calculateMonthDay();
+        // Get UTC date components
+        const year = year_and_day.year;
+        const day = @as(u32, month_and_day.day_index) + 1; // this starts in 0
+        const month = month_and_day.month.numeric(); // starts in 1
+
+        // Get UTC time components
+        const time = utc_seconds.getDaySeconds();
+        const hours = time.getHoursIntoDay();
+        const minutes = time.getMinutesIntoHour();
+        const seconds = time.getSecondsIntoMinute();
+
+        // Format the date
+        return .{
+            .numeric_day = secs - time.secs,
+            .date = std.fmt.allocPrint(allocator, "{d:0>4}{d:0>2}{d:0>2}T{d:0>2}{d:0>2}{d:0>2}Z", .{
+                year,
+                month,
+                day,
+                hours,
+                minutes,
+                seconds,
+            }) catch bun.outOfMemory(),
+        };
+    }
+
+    const DIGESTED_HMAC_256_LEN = 32;
+    threadlocal var SIGNATURE_CACHE: bun.StringArrayHashMap([DIGESTED_HMAC_256_LEN]u8) = undefined;
```

Comment:
this will leak memory  and the key will be invalid, we can probably get away with not doing this

Reply:
I still think we need signature cache but we can do it later in a better way